src/backend/distributed/commands/multi_copy.c

```diff
@@ -2048,8 +2048,12 @@ CitusCopyDestReceiverStartup(DestReceiver *dest, int operation,
 
 	UseCoordinatedTransaction();
 
-	/* all modifications use 2PC */
-	Use2PCForCoordinatedTransaction();
+	/* all modifications use 2PC unless we only touch a single shard */
+	if (!IsCitusTableTypeCacheEntry(cacheEntry, SINGLE_SHARD_DISTRIBUTED) &&
```

Comment:
Can't we check if `shardIntervalList` has more than 1 shards?

Reply:
Yeah, I also think that we should consider generalizing the code.

I think we should *probably* postpone the 2PC decision until `CitusCopyDestReceiverShutdown()`,  there iterate over `shardStateHash` and make sure that only a single node is involved  (CopyShardState->placementStateList->groupId).

Reply:
> probably

I said probably because I cannot exactly remember whether we ever clear/invalidate CopyShardState before copy finishes. If so, we cannot follow this suggestion

Reply:
we probably need similar checks as in Activate2PCIfModifyingTransactionExpandsToNewNode